teps/0091-trusted-resources.md

```diff
@@ -0,0 +1,335 @@
+this will be left out of the scope of this TEP.
+
+The Pipeline Author will be able to provide optional configuration to Tekton
+Chains such that Chains will *only* consider results from Tasks that were
+verifiable (and, following the previous rule, were successfully verified).
+
+
+### Risks and Mitigations
+
+* Mutating webhook is a risk of the proposal. The proposed verification will happen after the mutating webhook, so the content of a Resource may be mutated and fail the verification.
+
+  Some possible  mitigations include the following:
+
+  1. Update the Resources to add the missing mutating values (e.g. Use `SetDefaults`) everytime we bump the version of Tekton Pipeline.
+  2. Avoid mutating Tekton Resources or move the mutating after verification.
+
+* When verifying a Remote Resource that identifies a Task, the Task still has
+references to external, unverified resources - namely,
+the builder images. These builder images may have been compromised and the
+Task would still be considered verified.
+
+  Some possible mitigations include the following:
+
+  1. (Within scope of this proposal) For a Task to be considered verified, all
+    builder image references within it are suggested to be by-digest. The digest references
+    would form a low-effort way for the Task Author to signal that they have
+    verified or otherwise trust the builder images they are referencing.
+    For example, they may have manually verified the builder images using Cosign
+    or some other technique.
+  2. (Outside the scope of this proposal) Configuration is introduced into Tekton
+    to facilitate verification of the builder images themselves. Configuration
+    would also be required to cover off various combinations of bundle
+    verification and image verification - for example, do I allow verified
+    Task Bundles if the Task within the Bundle does not have verified images?
+
+### User Experience (optional)
+
+**Author publishing a Resource.**
+A Resource Author may want their Resource and its
+outputs to be trusted so that provenance attestations can be confidently made. The author will use the private key to sign the Resource and signatures are stored in `Annotations` map of the Resource. Public keys are stored as secrets or via URI stored in the configmap. Similar mechanism is used in [Chains](https://github.com/tektoncd/chains).
+
+**Verify the Resource via validating webhook.**
+The user will need to deploy the validating webhook to the cluster, by default the webhook will reject a Resource if it fails the verification. The webhook will mark if this Resource fails or passes the verification.
+
+**Configure the Chains to skip the run if failed verificaiton.**
+Chains can be configured to not create a provenance attestation if the verification fails.
+
+
+### Performance (optional)
+
+For remote Resources such as OCI bundle, it will take time to fetch the resources. Resources that are not
```

Comment:
Do you have a rough estimate of how long the resource fetch takes?

Reply:
I don't have right now. It will depend on how revolver work and also what remote source we are fetching. For now I can make some experiment, like for the OCI case, if I pull it from docker hub, how long it will take. To get a rough estimate of it!

Reply:
Tested a simple task oci bundle fetching from docker hub and the time is very short, can be ignored.

Reply:
I'm still a bit worried about the remote artifact fetch particularly if this is done as a validation webhook and uses remote resolution - that would involve the webhook creating a ResolutionRequest CR, waiting for the Resolver to finish resolving, and making a validation decision. 

(One mitigation would be to do this verification in the reconciler instead of in the admission webhook)

Reply:
Yes, and this also needs to give webhook the permission to fetch the tasks/pipelines. It's not so difficult to do it after resolution.

Reply:
Something to balance when thinking about webhook vs reconcile validation - validation before admission is the safer strategy because the resource can be blocked before downstream informers are aware of them.

Other OCI based policy webhooks (OPA, Kyverno, sigstore/policy-controller) seem to do fine with network based lookups at admission time, though I get we also want to dedup fetching efforts between this and RemoteResolution.

If we go down the reconcile validation path, we should be careful about documenting when resources are considered trusted and how that information is exposed to users in the Run status (e.g. it might be worth having a dedicated termination status marking a Run as unverified?). Even then, there may still be some utility in a webhook based validation to make sure that the Run object itself hasn't been tampered with. 🤔

Reply:
Agreed with the caveats around reconciler based validation but I do think there is (or will be) a difference in the remote fetch between this webhook and OPA/policy controller - with remote resolution for resource fetches, the webhook will have to create a custom resource and then watch for update to its status as opposed to a sync HTTP request/response